File: application-service/src/main/java/com/example/loanapp/service/RuleEvaluator.java
package com.example.loanapp.service;

import com.example.common.template.*;
import org.springframework.stereotype.Component;
import java.util.Map;

@Component
public class RuleEvaluator {

    public boolean isVisible(FieldDefinition field, Map<String, String> answers) {
        if (field.visibilityRules == null || field.visibilityRules.isEmpty()) return true;
        boolean hasShowRules = field.visibilityRules.stream().anyMatch(r -> r.show);
        if (hasShowRules) {
            return field.visibilityRules.stream()
                .filter(r -> r.show)
                .anyMatch(r -> evaluateCondition(r.condition, answers));
        }
        return field.visibilityRules.stream()
            .filter(r -> !r.show)
            .noneMatch(r -> evaluateCondition(r.condition, answers));
    }

    public boolean isRequired(FieldDefinition field, Map<String, String> answers) {
        if (field.required) return true;
        if (field.validationRules == null) return false;
        return field.validationRules.stream()
            .filter(r -> r.validationType == ValidationType.REQUIRED)
            .anyMatch(r -> r.condition == null || evaluateCondition(r.condition, answers));
    }

    public boolean evaluateCondition(RuleCondition condition, Map<String, String> answers) {
        if (condition == null) return true;
        String actual = answers.getOrDefault(condition.fieldKey, "");
        return switch (condition.operator) {
            case EQUALS       -> condition.value != null && condition.value.equals(actual);
            case NOT_EQUALS   -> condition.value != null && !condition.value.equals(actual);
            case IN           -> condition.values != null && condition.values.contains(actual);
            case NOT_IN       -> condition.values == null || !condition.values.contains(actual);
            case IS_EMPTY     -> actual.isBlank();
            case IS_NOT_EMPTY -> !actual.isBlank();
        };
    }
}

File: application-ui/src/ruleEvaluator.js
export function isVisible(field, answers) {
  const rules = field.visibilityRules || []
  if (rules.length === 0) return true
  const showRules = rules.filter(r => r.show)
  if (showRules.length > 0) {
    return showRules.some(r => evalCondition(r.condition, answers))
  }
  return rules.filter(r => !r.show).every(r => !evalCondition(r.condition, answers))
}

export function isRequired(field, answers) {
  if (field.required) return true
  return (field.validationRules || [])
    .filter(r => r.validationType === 'REQUIRED')
    .some(r => !r.condition || evalCondition(r.condition, answers))
}

function evalCondition(condition, answers) {
  if (!condition) return true
  const actual = answers[condition.fieldKey] || ''
  switch (condition.operator) {
    case 'EQUALS':       return condition.value === actual
    case 'NOT_EQUALS':   return condition.value !== actual
    case 'IN':           return (condition.values || []).includes(actual)
    case 'NOT_IN':       return !(condition.values || []).includes(actual)
    case 'IS_EMPTY':     return !actual
    case 'IS_NOT_EMPTY': return !!actual
    default:             return false
  }
}
